File: LeetCode/15.三数之和.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
		sort(nums.begin(), nums.end());

		vector<vector<int>> ans;
		for (int i = 0, cnt = nums.size(); i < cnt;){
			int cur = nums[i];
			this->twoSum(nums, -cur, i + 1, ans);
			while (++i < cnt && cur == nums[i]);
		}
		return ans;
    }

	// 如何避免返回值拷贝？
	void twoSum(vector<int>& nums, const int& tgt, const int& start, vector<vector<int>>& ans){
		int lo = start;
		int hi = nums.size() - 1;
		while (lo < hi){
			int tmp = nums[lo] + nums[hi];
			if (tmp < tgt){
				lo++;
			}else if (tmp > tgt){
				hi--;
			}else{
				int left = nums[lo];
				int right = nums[hi];
				ans.push_back({-tgt, left, right});

				while (++lo < hi && left == nums[lo]);
				while (lo < --hi && right == nums[hi]);
			}
		}
	}
};

File: LeetCode/15.三数之和.py
# -*- coding: utf-8 -*-
'''
给你一个包含 n 个整数的数组 nums，判断 nums 中是否存在三个元素 a，b，c ，使得 a + b + c = 0 ？请你找出所有和为 0 且不重复的三元组。

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/3sum
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
'''
class Solution(object):
	'''
	'''
	def threeSum(self, nums):
		"""
		:type nums: List[int]
		:rtype: List[List[int]]
		"""
		nums.sort()

		l = len(nums)
		i = 0

		ret = []
		while i < l:
			ret.extend(self.twoSum(nums, -nums[i], i + 1))
			old = nums[i]
			while i < l and nums[i] == old:
				i += 1
		return nums

	def twoSum(self, nums, tgt, start):
		lo, hi = start, len(nums) - 1
		ret = []
		while lo < hi:
			tmp = nums[lo] + nums[hi]
			if tmp < tgt:
				lo += 1
			elif tmp > tgt:
				hi -= 1
			else:
				left = nums[lo]
				right = nums[hi]
				ret.append([-tgt, left, right])
				while hi > lo and nums[lo] == left:
					lo += 1
				while hi > lo and nums[hi] == right:
					hi -= 1
		return ret

